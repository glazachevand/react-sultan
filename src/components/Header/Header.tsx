import React from 'react';
import logo from "./img/logo.png"
import call from "./img/call.png"
import { Link } from "react-router-dom";
import CartBtn from '../UI/CartBtn/CartBtn';

const Header: React.FC = () => {
  const [isOpen, setIsOpen] = React.useState(false)

  const onClickBurger = () => {
    setIsOpen(!isOpen);
  }

  return (
    <header className="header">
      <div className="header__menu-row">
        <div className="header__menu-row-container _container">
          <div className="header__address">
            <div className="_icon-location"></div>
            <div className="header__address-text">г. Кокчетав, ул. Ж. Ташенова 129Б<br /><span>(Рынок Восточный)</span></div>
          </div>
          <div className="header__mail">
            <div className="_icon-mail"></div>
            <div className="header__mail-text"><a href="mailto:">opt.sultan@mail.ru</a><br /><span>На связи в любое время</span></div>
          </div>
          <nav className="header__menu menu">
            <ul className="menu__list">
              <li><a href="#" className="menu__link">О компании</a></li>
              <li><a href="#" className="menu__link">Доставка и оплата</a></li>
              <li><a href="#" className="menu__link">Возврат</a></li>
              <li><a href="#" className="menu__link">Контакты</a></li>
            </ul>
          </nav>
        </div>
      </div>
      <div className="header__logo-row">
        <div className="header__logo-row-container _container">
          <Link to="/" className="header__logo"><img src={logo} alt="logo" /></Link>
          <a href="#" className="header__catalog btn yellow-btn">Каталог<svg width="15" height="15" viewBox="0 0 15 15" fill="none"
            xmlns="http://www.w3.org/2000/svg">
            <path
              d="M0.5 2C0.5 1.17157 1.17157 0.5 2 0.5H6C6.27614 0.5 6.5 0.723858 6.5 1V6C6.5 6.27614 6.27614 6.5 6 6.5H1C0.723858 6.5 0.5 6.27614 0.5 6V2Z"
              stroke="white" />
            <path
              d="M8.5 1C8.5 0.723858 8.72386 0.5 9 0.5H13C13.8284 0.5 14.5 1.17157 14.5 2V6C14.5 6.27614 14.2761 6.5 14 6.5H9C8.72386 6.5 8.5 6.27614 8.5 6V1Z"
              stroke="white" />
            <path
              d="M8.5 9C8.5 8.72386 8.72386 8.5 9 8.5H14C14.2761 8.5 14.5 8.72386 14.5 9V13C14.5 13.8284 13.8284 14.5 13 14.5H9C8.72386 14.5 8.5 14.2761 8.5 14V9Z"
              stroke="white" />
            <path
              d="M0.5 9C0.5 8.72386 0.723858 8.5 1 8.5H6C6.27614 8.5 6.5 8.72386 6.5 9V14C6.5 14.2761 6.27614 14.5 6 14.5H2C1.17157 14.5 0.5 13.8284 0.5 13V9Z"
              stroke="white" />
          </svg>
          </a>
          <div className="search  header__search">
            <form className="search__form" name="searchForm" action="#" method="get">
              <input type="search" className="search__input search-input" placeholder="Поиск..." />
              <button className="search__btn search-btn">
                <svg width="19" height="19" viewBox="0 0 19 19" fill="none" xmlns="http://www.w3.org/2000/svg">
                  <path
                    d="M16.5294 16.5294L13.0989 13.0928L16.5294 16.5294ZM15 8.5C15 10.2239 14.3152 11.8772 13.0962 13.0962C11.8772 14.3152 10.2239 15 8.5 15C6.77609 15 5.12279 14.3152 3.90381 13.0962C2.68482 11.8772 2 10.2239 2 8.5C2 6.77609 2.68482 5.12279 3.90381 3.90381C5.12279 2.68482 6.77609 2 8.5 2C10.2239 2 11.8772 2.68482 13.0962 3.90381C14.3152 5.12279 15 6.77609 15 8.5V8.5Z"
                    stroke="white" strokeWidth="1.3" strokeLinecap="round" />
                </svg>
              </button>
            </form>
          </div>
          <div className="header__tel-content">
            <div className="header__tel-text">
              <a href="tel:+77774900091" className="header__tel">+7 (777) 490-00-91</a>
              <div className="header__work-hours">время работы: 9:00-20:00</div>
              <a href="#" className="header__tel-request">Заказать звонок</a>
            </div>
            <div className="header__tel-img"><img src={call} alt="call" /></div>
          </div>
          <a href="#" className="header__price-list btn yellow-btn" download="">Прайс-лист <div className="_icon-download"></div>
          </a>
          <CartBtn cssClass={'header__cart'} />
        </div>
      </div>
      <div className="header__menu-row-lg">
        <div className="header__menu-row-lg-container _container">
          <div className={isOpen ? `icon-menu _open` : `icon-menu`} onClick={onClickBurger}>
            <div className="icon-menu__burger">
              <span></span>
              <span></span>
              <span></span>
            </div>
          </div>
          <a href="#!" className="header__logo"><img src={logo} alt="logo" /></a>
          <CartBtn cssClass={'header__cart'} />
        </div>
      </div>
      <div className="header__logo-row-lg">
        <div className="header__logo-row-lg-container _container">
          <a href="#" className="header__catalog btn yellow-btn">Каталог<svg viewBox="0 0 15 15" fill="none" xmlns="http://www.w3.org/2000/svg">
            <path
              d="M0.5 2C0.5 1.17157 1.17157 0.5 2 0.5H6C6.27614 0.5 6.5 0.723858 6.5 1V6C6.5 6.27614 6.27614 6.5 6 6.5H1C0.723858 6.5 0.5 6.27614 0.5 6V2Z" />
            <path
              d="M8.5 1C8.5 0.723858 8.72386 0.5 9 0.5H13C13.8284 0.5 14.5 1.17157 14.5 2V6C14.5 6.27614 14.2761 6.5 14 6.5H9C8.72386 6.5 8.5 6.27614 8.5 6V1Z" />
            <path
              d="M8.5 9C8.5 8.72386 8.72386 8.5 9 8.5H14C14.2761 8.5 14.5 8.72386 14.5 9V13C14.5 13.8284 13.8284 14.5 13 14.5H9C8.72386 14.5 8.5 14.2761 8.5 14V9Z" />
            <path
              d="M0.5 9C0.5 8.72386 0.723858 8.5 1 8.5H6C6.27614 8.5 6.5 8.72386 6.5 9V14C6.5 14.2761 6.27614 14.5 6 14.5H2C1.17157 14.5 0.5 13.8284 0.5 13V9Z" />
          </svg>
          </a>
          <div className="search  header__search">
            <form className="search__form" name="searchForm" action="#" method="get">
              <input type="search" className="search__input search-input" placeholder="Поиск..." />
              <button className="search__btn search-btn">
                <svg viewBox="0 0 19 19" fill="none" xmlns="http://www.w3.org/2000/svg">
                  <path
                    d="M16.5294 16.5294L13.0989 13.0928L16.5294 16.5294ZM15 8.5C15 10.2239 14.3152 11.8772 13.0962 13.0962C11.8772 14.3152 10.2239 15 8.5 15C6.77609 15 5.12279 14.3152 3.90381 13.0962C2.68482 11.8772 2 10.2239 2 8.5C2 6.77609 2.68482 5.12279 3.90381 3.90381C5.12279 2.68482 6.77609 2 8.5 2C10.2239 2 11.8772 2.68482 13.0962 3.90381C14.3152 5.12279 15 6.77609 15 8.5V8.5Z"
                    strokeWidth="1.3" strokeLinecap="round" />
                </svg>
              </button>
            </form>
            <div className="search__title">Поиск</div>
          </div>
        </div>
      </div>
      <div className={isOpen ? `burger-menu _open` : `burger-menu`}>
        <div className="burger-menu__body">
          <div className="_container">
            <div className="header__address">
              <div className="_icon-location"></div>
              <div className="header__address-text">г. Кокчетав, ул. Ж. Ташенова 129Б<br /><span>(Рынок Восточный)</span></div>
            </div>
            <div className="header__mail">
              <div className="_icon-mail"></div>
              <div className="header__mail-text"><a href="mailto:">opt.sultan@mail.ru</a><br /><span>На связи в любое время</span></div>
            </div>
            <div className="header__tel-content">
              <div className="_icon-tel"></div>
              <div className="header__tel-text">
                <h3 className="header__tel-title">Отдел продаж</h3>
                <a href="tel:+77774900091" className="header__tel">+7 (777) 490-00-91</a>
                <div className="header__work-hours">время работы: 9:00-20:00</div>
              </div>
            </div>
            <div className="header__request-container">
              <div className="_icon-tel2 header__icon-request"></div>
              <a href="#" className="header__tel-request">Заказать звонок</a>
            </div>
            <nav className="header__menu menu">
              <h1 className="menu__title">Меню сайта:</h1>
              <ul className="menu__list">
                <li><a href="#" className="menu__link">О компании</a></li>
                <li><a href="#" className="menu__link">Доставка и оплата</a></li>
                <li><a href="#" className="menu__link">Возврат</a></li>
                <li><a href="#" className="menu__link">Контакты</a></li>
              </ul>
            </nav>
            <a href="#" className="header__price-list btn yellow-btn" download="">Прайс-лист <div className="_icon-download"></div></a>
          </div>
        </div>
      </div>
    </header>
  );
};

export default Header;
