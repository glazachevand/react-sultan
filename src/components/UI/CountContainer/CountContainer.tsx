import React from 'react';
import { useActions } from '../../../hooks/useActions';
import { CartItemType } from '../../../types/cart';
import { useTypedSelector } from '../../../hooks/useTypedSelector';

interface CountContainerProps {
  id: number
}

const CountContainer: React.FC<CountContainerProps> = ({ id }) => {
  const { products } = useTypedSelector(state => state.productsRed);
  const product = products.find(item => item.id === Number(id));
  const { items } = useTypedSelector(state => state.cart);
  let cartItem = items.find(item => item.id === Number(id));

  if (product && !cartItem) {
    cartItem = {
      id: product.id, title: product.title, url: product.url,
      brand: product.brand, description: product.description, typesize: product.typesize,
      size: product.size, price: product.price, count: 0
    }
  }

  const { addCartItem, minusCartItem } = useActions();

  const onClickPlus: React.MouseEventHandler<HTMLButtonElement> = () => {
    addCartItem(cartItem as CartItemType);
  };

  const onClickMinus: React.MouseEventHandler<HTMLButtonElement> = () => {
    minusCartItem(cartItem as CartItemType);
  };

  return (
    <div className="cart__counts">
      <button className="minus" onClick={onClickMinus} disabled={cartItem?.count === 1}>-</button>
      <div className="count">{cartItem?.count || 0}</div>
      <button className="plus" onClick={onClickPlus}>+</button>
    </div>
  );
};

export default CountContainer;