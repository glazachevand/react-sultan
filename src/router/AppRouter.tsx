import React from 'react';
import { Routes, Route } from 'react-router-dom';
import Cart from '../pages/Cart/Cart';
import FullProduct from '../pages/FullProduct/FullProduct';
import Admin from '../pages/Admin/Admin';
import Catalog from '../pages/Catalog/Catalog';

const AppRouter: React.FC = () => {
  return (
    <Routes>
      <Route path="/" element={<Catalog />} />
      <Route path="/cart" element={<Cart />} />
      <Route path="/admin" element={<Admin />} />
      <Route path="/product/:id" element={<FullProduct />} />
      <Route path="*" element={<Catalog />} />
    </Routes>
  );
};

export default AppRouter;