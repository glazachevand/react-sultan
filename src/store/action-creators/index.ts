import * as ProductsActionCreators from './products';
import * as FilterActionCreators from './filter';
import * as CartActionCreators from './cart';

export default {
  ...ProductsActionCreators,
  ...FilterActionCreators,
  ...CartActionCreators,
};
