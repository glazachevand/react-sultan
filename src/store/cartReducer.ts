import { CartState, CartsAction, CartActionTypes } from '../types/cart';
import { calcTotalPrice } from '../utils/calcTotalPrice';
import { getCartFromLS } from '../utils/getCartFromLS';

const initialState: CartState = getCartFromLS();

export const cartReducer = (state = initialState, action: CartsAction): CartState => {
  let totalPrice, itemsCopy, findItem;

  switch (action.type) {
    case CartActionTypes.ADD_ITEM:
      itemsCopy = [...state.items];
      findItem = itemsCopy.find((item) => item.id === action.payload.id);

      if (findItem) {
        findItem.count++;
      } else {
        itemsCopy.push({ ...action.payload, count: 1 });
      }

      totalPrice = calcTotalPrice(itemsCopy);
      return { ...state, items: itemsCopy, totalPrice };

    case CartActionTypes.DELETE_ITEM:
      const itemsFilter = [...state.items.filter((item) => item.id !== action.payload.id)];
      totalPrice = calcTotalPrice(itemsFilter);
      return { ...state, items: itemsFilter, totalPrice };

    case CartActionTypes.MINUS_ITEM:
      itemsCopy = [...state.items];
      findItem = itemsCopy.find((item) => item.id === action.payload.id);

      if (findItem) {
        findItem.count--;
      }

      totalPrice = calcTotalPrice(itemsCopy);
      return { ...state, items: itemsCopy, totalPrice };

    case CartActionTypes.CLEAR_ITEMS:
      return { ...state, items: [], totalPrice: 0 };

    default:
      return state;
  }
};
