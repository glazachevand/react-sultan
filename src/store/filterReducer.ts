import { FilterState, FilterAction, FiltersActionTypes } from '../types/filter';
import { getManufactures } from '../utils/getManufactures';
import data from './../data/data.json';

const initialState: FilterState = {
  sort: 'price_asc',
  category: 'Все',
  priceMin: 10,
  priceMax: 10000,
  manufacturers: getManufactures(data.products),
  page: 1,
  filterProducts: data.products,
};

export const filterReducer = (state = initialState, action: FilterAction): FilterState => {
  let productsCopy;
  switch (action.type) {
    case FiltersActionTypes.SET_CATEGORY:
      return { ...state, category: action.payload };

    case FiltersActionTypes.SET_SORT:
      productsCopy = [...state.filterProducts];

      switch (action.payload) {
        case 'title_asc':
          productsCopy = productsCopy.sort((a, b) => `${a.brand} ${a.title}`.localeCompare(`${b.brand} ${b.title}`));
          break;

        case 'title_desc':
          productsCopy = productsCopy.sort((a, b) => `${b.brand} ${b.title}`.localeCompare(`${a.brand} ${a.title}`));
          break;

        case 'price_asc':
          productsCopy = productsCopy.sort((a, b) => a.price - b.price);
          break;

        case 'price_desc':
          productsCopy = productsCopy.sort((a, b) => b.price - a.price);
          break;

        default:
          break;
      }

      return { ...state, sort: action.payload, filterProducts: productsCopy };

    case FiltersActionTypes.SET_PAGE:
      return { ...state, page: action.payload };

    case FiltersActionTypes.FILTER_PRODUCTS:
      productsCopy = [...action.payload.products];

      if (action.payload.category !== 'Все') {
        productsCopy = productsCopy.filter((product) => product.typecare.includes(action.payload.category));
      }

      productsCopy = productsCopy.filter((product) => product.price < action.payload.priceMax && product.price > action.payload.priceMin);

      productsCopy = productsCopy.filter((product) => action.payload.manufacturers.includes(product.manufacturer));

      return { ...state, filterProducts: productsCopy };

    case FiltersActionTypes.SET_PARAMETERS:
      return { ...state, priceMin: action.payload.priceMin, priceMax: action.payload.priceMax, manufacturers: action.payload.manufacturers };

    case FiltersActionTypes.CLEAR_PARAMETERS:
      return { ...state, priceMin: 10, priceMax: 10000, manufacturers: action.payload };

    default:
      return state;
  }
};
