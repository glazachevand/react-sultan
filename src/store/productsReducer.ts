import { ProductsState, ProductsAction, ProductsActionTypes } from '../types/products';
import { getProductsFromLS } from './../utils/getProductsFromLS';

const initialState: ProductsState = getProductsFromLS();

export const productsReducer = (state = initialState, action: ProductsAction): ProductsState => {
  let arrCopy, index;
  switch (action.type) {
    case ProductsActionTypes.DELETE_PRODUCT:
      return { ...state, products: [...state.products.filter((product) => product.id !== action.payload.id)] };

    case ProductsActionTypes.ADD_PRODUCT:
      return { ...state, products: [...state.products, action.payload] };

    case ProductsActionTypes.UPDATE_PRODUCT:
      arrCopy = [...state.products];
      index = arrCopy.findIndex((item) => item.id === action.payload.id);
      arrCopy.splice(index, 1, action.payload);
      return { ...state, products: arrCopy };

    case ProductsActionTypes.DELETE_CATEGORY:
      arrCopy = [...state.categories].filter((item) => item !== action.payload);
      return { ...state, categories: arrCopy };

    case ProductsActionTypes.UPDATE_CATEGORIES:
      return { ...state, categories: action.payload };

    case ProductsActionTypes.ADD_CATEGORY:
      return { ...state, categories: [...state.categories, action.payload] };

    default:
      return state;
  }
};
