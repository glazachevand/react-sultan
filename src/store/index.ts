import { createStore, combineReducers } from 'redux';
import { productsReducer } from './productsReducer';
import { filterReducer } from './filterReducer';
import { composeWithDevTools } from '@redux-devtools/extension';
import { cartReducer } from './cartReducer';

export const rootReducer = combineReducers({
  productsRed: productsReducer,
  filter: filterReducer,
  cart: cartReducer,
});

export const store = createStore(rootReducer, composeWithDevTools());
export type RootState = ReturnType<typeof rootReducer>;
