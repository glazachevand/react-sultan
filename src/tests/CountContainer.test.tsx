import { render, screen, fireEvent } from '@testing-library/react'
import { Provider } from "react-redux";
import { MemoryRouter } from 'react-router-dom';
import { rootReducer } from '../store';
import { createStore, Store } from 'redux';
import CountContainer from '../components/UI/CountContainer/CountContainer';
import AppRouter from '../router/AppRouter';
import Header from '../components/Header/Header';

describe('Test counter container', () => {
  let initialState: ReturnType<typeof store.getState>;
  let store: Store;

  beforeEach(() => {
    initialState = {
      productsRed: {
        products: [
          {
            id: 1,
            title: 'BIO-SOAP Экологичное туалетное мыло. Литсея и бергамот',
            url: 'product1.webp',
            barcode: 4604049097548,
            manufacturer: 'BioMio',
            brand: 'BioMio',
            description: 'Экологичное гипоаллергенное туалетное мыло BioMio.',
            typesize: 'вес',
            size: '1020 г',
            typecare: [
              'Уход за руками',
              'Уход за телом',
              'Уход за ногами'
            ],
            price: 95
          },
        ],
        categories: [
          'Все',
          'Уход за телом',
          'Уход за руками',
          'Уход за ногами',
          'Уход за лицом',
          'Уход за волосами',
          'Средства для загара',
          'Средства для бритья',
          'Подарочные наборы',
          'Гигиеническая продукция',
          'Гигиена полости рта',
          'Бумажная продукция'
        ]
      },
      filter: {
      },
      cart: {
        items: [
          {
            id: 1,
            title: 'BIO-SOAP Экологичное туалетное мыло. Литсея и бергамот',
            url: 'product1.webp',
            brand: 'BioMio',
            description: 'Экологичное гипоаллергенное туалетное мыло BioMio.',
            typesize: 'вес',
            size: '1020 г',
            price: 95,
            count: 2
          }
        ],
        totalPrice: 190
      }
    }

    store = createStore(rootReducer, initialState);
  });

  test('Click on buttons - check counts', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/cart']}>
          <CountContainer id={1} />
        </MemoryRouter>
      </Provider>
    );

    const minus = screen.getByTestId('minus');
    const plus = screen.getByTestId('plus');
    expect(screen.getByTestId('count')).toHaveTextContent('2')
    fireEvent.click(plus);
    expect(screen.getByTestId('count')).toHaveTextContent('3')
    fireEvent.click(plus);
    expect(screen.getByTestId('count')).toHaveTextContent('4')
    fireEvent.click(minus);
    expect(screen.getByTestId('count')).toHaveTextContent('3')
    expect(screen.getByTestId('count')).toMatchSnapshot();
  });

  test('Click on buttons plus - check cart in the header', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/cart']}>
          <Header />
          <main className="main">
            <AppRouter />
          </main>
        </MemoryRouter>
      </Provider>
    );

    const plus = screen.getByTestId('plus');
    expect(screen.getByTestId('count')).toHaveTextContent('2')
    expect(screen.getAllByTestId('cart-icon-count')[0]).toHaveTextContent('2')
    fireEvent.click(plus);
    expect(screen.getByTestId('count')).toHaveTextContent('3')
    expect(screen.getAllByTestId('cart-icon-count')[0]).toHaveTextContent('3')
    fireEvent.click(plus);
    expect(screen.getByTestId('count')).toHaveTextContent('4')
    expect(screen.getAllByTestId('cart-icon-count')[0]).toHaveTextContent('4')
  });
});