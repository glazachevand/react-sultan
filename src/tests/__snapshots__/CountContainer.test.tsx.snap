// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test counter container Click on buttons - check counts 1`] = `
<div
  class="count"
  data-testid="count"
>
  3
</div>
`;

exports[`Test counter container Click on buttons 1`] = `
<div
  class="count"
  data-testid="count"
>
  3
</div>
`;
